### PR TITLE
Token authentication supprot for buffer server

### PR DESCRIPTION
**Pull request by PramodSSImmaneni**
Token authentication for buffer server to authenticate clients before allowing them. Each buffer server is assigned its own token and the clients use the appropriate token when talking to the corresponding server(s). This authentication is enabled only in secure mode and was tested on it.

### CODE REVIEW

File: bufferserver/src/main/java/com/datatorrent/bufferserver/client/AuthClient.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        int match = 0;
+        while ((match < token.length) && (buffer[offset + match] == token[match])) {
+          ++match;
+        }
+        if (match == token.length) {
+          authenticated = true;
+        }
+      }
+      if (!authenticated) {
+        throw new RuntimeException("Authentication failure");
```
Comment (243826) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/client/AuthClient.java]: Can it be an AccessControlException instead, with a more discreet message?

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 2]
```diff
@@ -103,6 +102,9 @@
 import com.datatorrent.stram.util.WebServicesClient;
 import com.datatorrent.stram.webapp.*;
 
+import net.engio.mbassy.bus.MBassador;
+import net.engio.mbassy.bus.config.BusConfiguration;
```
Comment (243826) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: .net goes back up, probably where it was.

File: engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java [hunk 3]
```diff
@@ -185,6 +185,9 @@ public void setup(StreamingContainerContext ctx)
         }
         // start buffer server, if it was not set externally
         bufferServer = new Server(0, blocksize * 1024 * 1024, blockCount);
+        if (UserGroupInformation.isSecurityEnabled()) {
```
Comment (243826) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: shouldn't this check be central and the context should be populated with token only if security enabled? Here the token should be set regardless from the context?

File: bufferserver/src/main/java/com/datatorrent/bufferserver/client/AuthClient.java (hunk 1 above)
Comment (PramodSSImmaneni) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/client/AuthClient.java]: Yes that's better

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java (hunk 2 above)
Comment (PramodSSImmaneni) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: Moved it above datatorrent imports

File: engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java (hunk 3 above)
Comment (PramodSSImmaneni) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: Yes was being extra careful and it just makes the decision distributed. Changed it.

**Comment by PramodSSImmaneni** [comment]
Addressed the comments

**Comment by PramodSSImmaneni** [comment]
Changed auth failure message. Let me know if it looks ok and I will create a new squash pull

File: bufferserver/src/main/java/com/datatorrent/bufferserver/client/AuthClient.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        int match = 0;
+        while ((match < token.length) && (buffer[offset + match] == token[match])) {
+          ++match;
+        }
+        if (match == token.length) {
+          authenticated = true;
+        }
+      }
+      if (!authenticated) {
+        throw new AccessControlException("Buffer server security is enabled. Access is restricted without proper credentials");
```
Comment (243826) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/client/AuthClient.java]: Looks good. A full stop will be better at the end :-)

**Comment by PramodSSImmaneni** [comment]
Closing this and opening a new squashed pull request with these additional changes

**Pull request closed without merging**